### DISCUSSION TITLE
Automatically generated share images

### DISCUSSION THREAD

**Discussion by jmduke** [buttondown/discussions#4] [Announcements]
## TL;DR

Buttondown will spin up a number of one-off OG image templates for each email that can be chosen when you send out emails. 

## What's the problem?

People love share cards. Or OG images. Or whatever you want to call them. The data [backs it up](https://www.agorapulse.com/social-media-lab/link-tweet-with-an-image-vs-tweet-with-twitter-cards-does-it-really-matter/):

> So while less people saw the tweets with the Twitter Cards more of them actually engaged with the tweets.

## What's the solution?

At a high level: this!

![image](https://user-images.githubusercontent.com/461046/212093454-5e51e3dd-64a8-4bea-aac6-afdac533776d.png)

Expanding the share image selector within the 'fiddly bits' modal to contain, instead of just a carousel, three options:

- Share image templates (better name to come, I'm sure) that get autogenerated
- Existing images on the email itself
- Custom images that you can upload

Those templates are pretty naive: they take metadata from the email (subject, author, icon, publish date) and render it using `vercel/@og`. The current plan is to launch with two templates:

- `subject-first`, which highlights the subject
- `author-first`, which highlights the author

## Rough timeline

I'm doing a good amount of the pre-work now, but the rollout is going to be in **February** once the existing product work on the RSS automation is in a good place.

## Open questions

- [ ]  What's the right set of 'templates' to launch with? Right now I think being pretty opinionated and (frankly) vanilla is good enough, so long as its easy to extend the existing templates over time

## Out of scope 

- As of _right this second_, this won't be for subscribe pages or archive pages or any of the non-email pages. That will probably change; it should be pretty easy to piggyback this generation logic.

**Comment by JayCuthrell** [comment]
I've had mixed results with "Customize fiddly bits" 

Primarily, I share on LinkedIn and the [Post Inspector](https://www.linkedin.com/post-inspector/) can seem arbitrary or default. 

Example: LinkedIn Post Inspector of my latest issue [Deploy Everything You Touch](https://www.linkedin.com/post-inspector/inspect/https:%2F%2Ffudge.org%2Farchive%2Fdeploy-everything-you-touch%2F)

`<meta name="image" property="og:image" content="[Image URL here]">` isn't satisfied and the result is default (apparently).

Result: 
![](https://buttondown-attachments.s3.us-west-2.amazonaws.com/images/5710841f-86ba-4df5-b019-121ad4f5a8e5.png)

Desired: (via "Customize fiddly bits")
<img width="607" alt="Screenshot 2023-01-22 at 11 19 47 PM" src="https://user-images.githubusercontent.com/95317/213965308-c19f8f23-82bd-4102-84b9-76f5711f4e1a.png">

**Reply by jmduke** [reply]
This was a weird one — it looks like LinkedIn might be grabbing the value from the `oembed` url rather than the `og:image`. I think I should be able to fix this today.

**Reply by JayCuthrell** [reply]
For example, the output for [Deploy Everything You Touch](https://fudge.org/archive/deploy-everything-you-touch/) is selecting the correct content but the meta tags are not what LinkedIn appears to be expecting.

`<meta name="og:image" content="https://buttondown-attachments.s3.us-west-2.amazonaws.com/images/487c68c1-fb59-4b8e-9d9a-d9565deb1cbb.png" />`

Based on the inspector feedback page, LinkedIn parses looking for this format: `<meta name="image" property="og:image" content="[Image URL here]">`

i.e.

`<meta name="image" property="og:image" content="https://buttondown-attachments.s3.us-west-2.amazonaws.com/images/487c68c1-fb59-4b8e-9d9a-d9565deb1cbb.png" />`

**Reply by jmduke** [reply]
Oh, fascinating. I will try tweaking those tags tomorrow morning; thanks for the flag.

**Reply by jmduke** [reply]
@JayCuthrell your sleuthing was correct! I'm seeing that the correct image was pulled in; can you verify on your end?

**Reply by JayCuthrell** [reply]
@jmduke Perfecto! 🙏

<img width="1168" alt="Screenshot 2023-01-30 at 1 44 56 PM" src="https://user-images.githubusercontent.com/95317/215566955-8b33cac7-ba44-4861-837f-b305503e76a4.png">

**Reply by JayCuthrell** [reply]
I'm not sure why but I believe the issue has regressed. 
<img width="1148" alt="Screenshot 2023-02-19 at 8 20 50 PM" src="https://user-images.githubusercontent.com/95317/219988540-d1675d50-1d2f-425a-b542-066119d22938.png">

**Reply by benborgers** [reply]
@JayCuthrell hm! It looks like LinkedIn somehow got to scraping this URL before the share image was added, and instead took the first image on the page (the profile photo). 

When I used a slightly different URL to circumvent their cache, it seems to work: 

![2023-02-20 at 08 55 18@2x](https://user-images.githubusercontent.com/30215449/220126821-4d72cf5a-d202-4bc1-82f0-213cb3e1a92c.png)

And now if I go back to using the normal URL, that seems to work as well? Maybe requesting the slightly different URL caused them to clear the cache for the base URL. 

![2023-02-20 at 08 56 34@2x](https://user-images.githubusercontent.com/30215449/220127128-9a7a633e-d1f0-45fa-92bd-b4578364bfe7.png)

**Reply by JayCuthrell** [reply]
You are correct. I think my workflow needs to add a delay before I share to LinkedIn due to that cache expiry.

**Reply by JayCuthrell** [reply]
Update: I have been doing edits on older posts and wanted to confirm you are even more correct. 😄 

LinkedIn appear to be very cache heavy on the first URL but treats any URI stem variation as a new cache object. So `/?foo` and `/?bar` are handled as different cache objects -- and the work you've done for oEmbed is working exactly as expected. Thank you again. 💯

**Reply by caseywatts** [reply]
I had totally written off the Linkedin post inspector -- so glad to know about this weird behavior, ty!

**Comment by jeremycherfas** [comment]
Quick comment; I've had ongoing problems with the Fiddly Bits, most notably even when there are two images in the email, I am offered only one to choose. I hope this feature will in future show all images. Thanks

**Reply by jmduke** [reply]
@jeremycherfas mind sharing with me a subject of an email where you were seeing that? Would love to dive in!

**Comment by jeremycherfas** [comment]
Latest one https://jbdpbt92.r.us-east-1.awstrack.me/L0/https:%2F%2Fbuttondown.email%2Fjeremycherfas%2Fsubscribers%2F6637ccb4-8ed5-4eba-bfa0-dac02201936f%2Farchive%2Feat-this-newsletter-197-more-or-less/1/01000186028f59ce-0d626752-e44c-4833-96d4-f829beaecdf9-000000/VFdhy4aWVcontDS8Vwe_jvOlk7I=306

I was shown only the graph, not the potato chips!

Jeremy

**Reply by jmduke** [reply]
Hmm. Are you still seeing that now (if you were to go back into that email?) Just checked on my end and I'm seeing the potato chips, so I'm wondering if there's some client-side parsing issue.

![image](https://user-images.githubusercontent.com/461046/216121725-8f72d265-9f20-4baa-838c-8c2170a870ed.png)

**Comment by jeremycherfas** [comment]
Interesting

On your screenshot, both images are adjacent. On mine, only one is visible, and although I clicked in the space there was nothing visible to indicate I should scroll.

I suppose if I had clicked on the image the green border might have given me a clue, but thinking about it, I don't believe that border showed up when I was clicking about on Sunday. :)

**Reply by jmduke** [reply]
Ha! So to make sure I understand correctly, you _do_ see the second image but it was hidden from view and not obvious to be scrollable? That's excellent feedback; I can make that behavior more visible.

**Comment by jeremycherfas** [comment]
Precisely.
